File: Problems/0415. Add Strings/Solution.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    string addStrings(string num1, string num2) {
        if (num2.size() > num1.size()) return addStrings(num2, num1);

        reverse(num1.begin(), num1.end());
        reverse(num2.begin(), num2.end());

        string ans;
        int carry = 0;
        for (int i = 0; i < num1.size(); i++) {
            int num_int1 = num1[i] - '0';
            int num_int2 = i < num2.size() ? num2[i] - '0' : 0;

            carry = num_int1 + num_int2 + carry;
            ans.push_back(char((carry % 10) + '0'));
            carry /= 10;
        }

        if (carry > 0) ans.push_back(char(carry + '0'));

        reverse(ans.begin(), ans.end());
        return ans;
    }
};

File: Problems/0415. Add Strings/Solution.rs
impl Solution {
    pub fn add_strings(num1: String, num2: String) -> String {
        if (num2.len() > num1.len()) {
            return Self::add_strings(num2, num1);
        }

        let mut rev_num1 = num1.chars().rev().collect::<String>();
        let mut rev_num2 = num2.chars().rev().collect::<String>();

        let mut ans = Vec::new();
        let mut carry:u8 = 0;
        for (i, c) in rev_num1.char_indices() {
            let int1 = rev_num1.as_bytes()[i] - b'0';
            let int2 = if i < rev_num2.len() { rev_num2.as_bytes()[i] - b'0' } else { 0 };

            carry = int1 + int2 + carry;
            ans.push(((carry % 10) + b'0') as char);
            carry = carry / 10;
        }

        if carry > 0 {
            ans.push((carry + b'0') as char);
        }

        ans.reverse();
        return ans.iter().collect();
    }
}